File: src/main/java/org/anhcraft/enc/utils/RouletteSelect.java
package org.anhcraft.enc.utils;

import org.anhcraft.spaciouslib.utils.RandomUtils;

import java.util.List;

/**
 * @see <a href="https://en.wikipedia.org/wiki/Fitness_proportionate_selection">https://en.wikipedia.org/wiki/Fitness_proportionate_selection</a>
 */
public class RouletteSelect {
    public static int chooseFromList(List<Double> list){
        double weight_sum = list.stream().mapToDouble(Double::doubleValue).sum();
        double value = RandomUtils.randomDouble(0, weight_sum);
        for(int i = 0; i < list.size(); i++) {
            if(value > 0) {
                value -= list.get(i);
            } else {
                return i;
            }
        }
        return list.size()-1;
    }
}

File: src/main/kotlin/dev.anhcraft.enc/utils/RouletteSelect.kt
package dev.anhcraft.enc.utils

import dev.anhcraft.jvmkit.utils.RandomUtil

/**
 * @see [https://en.wikipedia.org/wiki/Fitness_proportionate_selection](https://en.wikipedia.org/wiki/Fitness_proportionate_selection)
 */
object RouletteSelect {
    @JvmStatic
    fun chooseFromList(list: List<Double>): Int {
        val weightsum = list.stream().mapToDouble{it.toDouble()}.sum()
        var value = RandomUtil.randomDouble(0.0, weightsum)
        for (i in list.indices) {
            if (value > 0) value -= list[i]
            else return i
        }
        return list.size - 1
    }
}
